### PR TITLE
feat: handle paging responses from apis

### LINKED ISSUE #764: feat: pagination change
## Motivation

@cchoirat 

## Proposal

Change the pagination from "1|2|3|4|5|Next" to a format that includes total page count, e.g. "Prev|1|2|...|[total pages - 1]|[total pages]|Next"

## Context

current:
image (1)

### LINKED ISSUE #1037: Handle paging correctly when fetching branches and commits
## Motivation

Currently, the functions to fetch commits and branches ignore the pagination headers and fetch the first 100 items (`getCommits` and `getBranches` in "src/api-client/repository.js").
This is creating problems for big projects where branches and commits are missing without making the user aware of it.
E.G. When trying to start a new environment in https://renkulab.io/projects/covid-19/covid-19-public-data/environments/new , the `master` branch is currently not available since the many `auto-update-xxx` branches are taking up all the space.

## Proposal

The easiest solution would be fetching all the pages when the pagination headers tell us there is more to come and consider the action unfinished until we get all the items.

In the long term, it would be nice to have a common mechanism to properly handle pagination when invoking the API. Currently, for these "array of objects" responses we query the API and expect a single monolithic response, setting the values `fetched, fetching` to make the presentational components aware of the current state. We should consider fetching a few items early and start displaying them while fetching the others.
Please note that a common strategy is not trivial here: we may want to throw away all the previous objects in some case, while in others we prefer keeping them until everything new has been fetched.

### PR DESCRIPTION
**Pull request by lorenzo-cavazzi**
This went a bit further than expected, introducing a couple of related (minor) changes.

* Introduces a method to fetch from API returning paginated data.
* Handles up to 1000 branches and commits (the limit is in place to prevent endless fetching loops that will fail anyway due to the gateway rate limit).
* Updates visualization where needed (project pages "overview --> commits" and "environments --> new")
* Minor changes to the Pagination component.

fix #1037, accidentally fix #764

P.S. All my namespaces are currently full so no preview available. You can run it on telepresence or contact me and I'll create a temporary preview.

### CODE REVIEW

**Review by ciyer — CHANGES_REQUESTED** [review]
Code looks good! I like the solution using an async iterator a lot; very elegant.

I have a couple of requested changes due to usability. I guess they are more requested-do-not-change-and-leave-as-before. I do not think a dropdown list with 100s of options is necessary because it just will not be useful, so I think we can leave the limits as they were before.

File: src/api-client/index.js [hunk 1]
```diff
@@ -133,6 +143,45 @@ class APIClient {
...
 
+  /**
+   * Create an iterable object to manage pagination. At each iteration, another page is fetched.
+   * You can use the for await syntax to fetch all the avilable pages
+   * E.G. for await (const partialData of clientIterableFetch(url)) { console.log(partialData) }
+   *
+   * @param {string} url - API url
+   * @param {object} [parameters] - Optional parameters object to be provided to clientFetch.
+   * @param {number} [parameters.maxIterations] - maximum iterations before throwing an error.
+   *   Default is 10. Set 0 for unlimited.
```
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply src/api-client/index.js]: I rephrased the comment a bit, let me know if it should be further improved

File: src/notebooks/Notebooks.present.js [hunk 2]
```diff
@@ -1049,10 +1049,10 @@ class StartNotebookCommitsOptions extends Component {
         <PopoverBody>
           <FormGroup>
             <Label>Number of commits to display</Label>
-            <Input type="number" min={0} max={100} step={1}
+            <Input type="number" min={0} max={1000} step={1}
```
[the remark this answers is not in the document]
Comment (lorenzo-cavazzi) [reply src/notebooks/Notebooks.present.js]: This makes sense, clogging up the dropdown menu won't be of much benefit.
I left the changes in the "Overview --> Commits" page untouched and I reverted the "Environment --> new" maximum commits to 100

**Review by ciyer — APPROVED** [review]
👍 +1

**Pull request merged**